[Bugfix]: make most of `test_openai_schema.py` pass

davidxia opened the pull request:
This change makes most of the parameterized tests in `pytest
tests/entrypoints/openai/test_openai_schema.py` pass. We handle user input errors, enrich HTTP error messages with more information, and update the error response definitions for several API endpoints.

The remaining tests that fail are `POST /tokenize` and maybe `POST
/v1/chat/completions`.
    
FIX #17037
FIX #17038

<details>
<summary>test results before</summary>

```bash
$ pytest tests/entrypoints/openai/test_openai_schema.py

...

===================================================================================== short test summary info =====================================================================================
(verbose_name='POST /tokenize') SUBFAIL tests/entrypoints/openai/test_openai_schema.py::test_openapi_stateless[POST /tokenize] - schemathesis.exceptions.CheckFailed: 
(verbose_name='POST /detokenize') SUBFAIL tests/entrypoints/openai/test_openai_schema.py::test_openapi_stateless[POST /detokenize] - schemathesis.exceptions.CheckFailed: 
(verbose_name='POST /v1/chat/completions') SUBFAIL tests/entrypoints/openai/test_openai_schema.py::test_openapi_stateless[POST /v1/chat/completions] - schemathesis.exceptions.CheckFailed: 
(verbose_name='POST /v1/completions') SUBFAIL tests/entrypoints/openai/test_openai_schema.py::test_openapi_stateless[POST /v1/completions] - schemathesis.exceptions.CheckFailed: 
(verbose_name='POST /v1/audio/transcriptions') SUBFAIL tests/entrypoints/openai/test_openai_schema.py::test_openapi_stateless[POST /v1/audio/transcriptions] - schemathesis.exceptions.CheckFailed: 
(verbose_name='POST /invocations') SUBFAIL tests/entrypoints/openai/test_openai_schema.py::test_openapi_stateless[POST /invocations] - schemathesis.exceptions.CheckFailed: 
(verbose_name='GET /metrics') SUBFAIL tests/entrypoints/openai/test_openai_schema.py::test_openapi_stateless[GET /metrics] - schemathesis.exceptions.CheckFailed: 
============================================================= 7 failed, 1 passed, 9 warnings, 14 subtests passed in 95.93s (0:01:35) ==============================================================
```
</details>


<details>
<summary>test results after</summary>

```bash
$ pytest tests/entrypoints/openai/test_openai_schema.py

...

===================================================================================== short test summary info =====================================================================================
(verbose_name='POST /tokenize') SUBFAIL tests/entrypoints/openai/test_openai_schema.py::test_openapi_stateless[POST /tokenize] - schemathesis.exceptions.CheckFailed: Schemathesis found 2 distinct sets of failures.
(verbose_name='POST /v1/chat/completions') SUBFAIL tests/entrypoints/openai/test_openai_schema.py::test_openapi_stateless[POST /v1/chat/completions] - schemathesis.exceptions.CheckFailed: Schemathesis found 0 distinct sets of failures.
============================================================= 2 failed, 1 passed, 9 warnings, 19 subtests passed in 267.55s (0:04:27) =============================================================
```
</details>

## 1. POST /tokenize

<details>
<summary>before</summary>

```bash
$ curl -X POST -H 'Content-Type: application/json' -d '{"messages": []}' --insecure http://127.0.0.1:8000/tokenize -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /tokenize HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 16
> 
< HTTP/1.1 400 Bad Request
< date: Mon, 05 May 2025 15:27:25 GMT
< server: uvicorn
< content-length: 80
< content-type: application/json
< 
* Connection #0 to host 127.0.0.1 left intact
{"object":"error","message":"","type":"BadRequestError","param":null,"code":400}
```
</details>

<details>
<summary>after</summary>

```bash
$ curl -X POST -H 'Content-Type: application/json' -d '{"messages": []}' --insecure http://127.0.0.1:8000/tokenize -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /tokenize HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 16
> 
< HTTP/1.1 400 Bad Request
< date: Mon, 12 May 2025 16:40:57 GMT
< server: uvicorn
< content-length: 104
< content-type: application/json
< 
* Connection #0 to host 127.0.0.1 left intact
{"object":"error","message":" list index out of range","type":"BadRequestError","param":null,"code":400}
```
</details>


## 2. POST /detokenize

<details>
<summary>before</summary>

```bash
$ curl -X POST -H 'Content-Type: application/json' -d '{"tokens": [-1]}' --insecure http://127.0.0.1:8000/detokenize -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /detokenize HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 16
> 
< HTTP/1.1 500 Internal Server Error
< date: Mon, 05 May 2025 15:27:50 GMT
< server: uvicorn
< content-length: 21
< content-type: text/plain; charset=utf-8
< 
* Connection #0 to host 127.0.0.1 left intact
Internal Server Error
```
</details>

<details>
<summary>after</summary>

```bash
$ curl -X POST -H 'Content-Type: application/json' -d '{"tokens": [-1]}' --insecure http://127.0.0.1:8000/detokenize -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /detokenize HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 16
> 
< HTTP/1.1 400 Bad Request
< date: Mon, 12 May 2025 16:41:46 GMT
< server: uvicorn
< content-length: 128
< content-type: application/json
< 
* Connection #0 to host 127.0.0.1 left intact
{"object":"error","message":" ['out of range integral type conversion attempted']","type":"Bad Request","param":null,"code":400}
```
</details>

## 3. POST /v1/chat/completions


<details>
<summary>before</summary>

```bash
$ curl -X POST -H 'Content-Type: application/json' -d '{"messages": []}' --insecure http://127.0.0.1:8000/v1/chat/completions -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /v1/chat/completions HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 16
> 
< HTTP/1.1 400 Bad Request
< date: Mon, 05 May 2025 15:28:16 GMT
< server: uvicorn
< content-length: 80
< content-type: application/json
< 
* Connection #0 to host 127.0.0.1 left intact
{"object":"error","message":"","type":"BadRequestError","param":null,"code":400}
```
</details>


<details>
<summary>after</summary>

```bash
$ curl -X POST -H 'Content-Type: application/json' -d '{"messages": []}' --insecure http://127.0.0.1:8000/v1/chat/completions -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /v1/chat/completions HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 16
> 
< HTTP/1.1 400 Bad Request
< date: Mon, 12 May 2025 16:42:10 GMT
< server: uvicorn
< content-length: 104
< content-type: application/json
< 
* Connection #0 to host 127.0.0.1 left intact
{"object":"error","message":" list index out of range","type":"BadRequestError","param":null,"code":400}
```
</details>

## 4. POST /v1/completions


<details>
<summary>before</summary>

```bash
$ curl -X POST -H 'Content-Type: application/json' -d '{"prompt": [-1]}' --insecure http://127.0.0.1:8000/v1/completions -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /v1/completions HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 16
> 
< HTTP/1.1 500 Internal Server Error
< date: Mon, 05 May 2025 15:35:48 GMT
< server: uvicorn
< content-length: 21
< content-type: text/plain; charset=utf-8
< 
* Connection #0 to host 127.0.0.1 left intact
Internal Server Error
```
</details>

<details>
<summary>after</summary>

```bash
$ curl -X POST -H 'Content-Type: application/json' -d '{"prompt": [-1]}' --insecure http://127.0.0.1:8000/v1/completions -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /v1/completions HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 16
> 
< HTTP/1.1 400 Bad Request
< date: Mon, 12 May 2025 16:43:17 GMT
< server: uvicorn
< content-length: 123
< content-type: application/json
< 
* Connection #0 to host 127.0.0.1 left intact
{"object":"error","message":"out of range integral type conversion attempted","type":"Bad Request","param":null,"code":400}
```
</details>

## 5. POST /v1/audio/transcriptions


<details>
<summary>before</summary>

```bash
$ curl -X POST -H 'Content-Type: application/x-www-form-urlencoded' -d file=b%27%27 --insecure http://127.0.0.1:8000/v1/audio/
transcriptions -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /v1/audio/transcriptions HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/x-www-form-urlencoded
> Content-Length: 12
> 
< HTTP/1.1 422 Unprocessable Entity
< date: Mon, 12 May 2025 16:48:48 GMT
< server: uvicorn
< content-length: 65
< content-type: application/json
< 
* Connection #0 to host 127.0.0.1 left intact
{"detail":"Expected 'file' to be a file-like object, not 'str'."}
```
</details>

<details>
<summary>after</summary>

```bash
$ curl -X POST -H 'Content-Type: application/x-www-form-urlencoded' -d file=b%27%27 --insecure http://127.0.0.1:8000/v1/audio/transcriptions -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /v1/audio/transcriptions HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/x-www-form-urlencoded
> Content-Length: 12
> 
< HTTP/1.1 422 Unprocessable Entity
< date: Mon, 12 May 2025 16:43:51 GMT
< server: uvicorn
< content-length: 137
< content-type: application/json
< 
* Connection #0 to host 127.0.0.1 left intact
{"object":"error","message":"Expected 'file' to be a file-like object, not 'str'.","type":"Unprocessable Entity","param":null,"code":422}
```
</details>

## 6. POST /invocations


<details>
<summary>before</summary>

```bash
$ curl -X POST --insecure http://localhost:8000/invocations -v
* Host localhost:8000 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:8000...
* connect to ::1 port 8000 from ::1 port 48744 failed: Connection refused
*   Trying 127.0.0.1:8000...
* Connected to localhost (127.0.0.1) port 8000
> POST /invocations HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/8.5.0
> Accept: */*
> 
< HTTP/1.1 415 Unsupported Media Type
< date: Mon, 05 May 2025 21:01:30 GMT
< server: uvicorn
< content-length: 71
< content-type: application/json
< 
* Connection #0 to host localhost left intact
{"detail":"Unsupported Media Type: Only 'application/json' is allowed"}

$ curl -X POST -H 'Content-Type: application/json' --insecure http://127.0.0.1:8000/invocations -v
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /invocations HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> 
< HTTP/1.1 500 Internal Server Error
< date: Mon, 05 May 2025 15:29:13 GMT
< server: uvicorn
< content-length: 21
< content-type: text/plain; charset=utf-8
< 
* Connection #0 to host 127.0.0.1 left intact
Internal Server Error
```
</details>

<details>
<summary>after</summary>

```bash
$ curl -X POST --insecure http://localhost:8000/invocations -v
* Host localhost:8000 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:8000...
* connect to ::1 port 8000 from ::1 port 35996 failed: Connection refused
*   Trying 127.0.0.1:8000...
* Connected to localhost (127.0.0.1) port 8000
> POST /invocations HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/8.5.0
> Accept: */*
> 
< HTTP/1.1 400 Bad Request
< date: Mon, 12 May 2025 16:44:29 GMT
< server: uvicorn
< content-length: 141
< content-type: application/json
< 
* Connection #0 to host localhost left intact
{"object":"error","message":" [\"Unsupported Media Type: Only 'application/json' is allowed\"]","type":"Bad Request","param":null,"code":400}

$ curl -X POST -H 'Content-Type: application/json' --insecure http://127.0.0.1:8000/invocations -v
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /invocations HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> 
< HTTP/1.1 400 Bad Request
< date: Mon, 12 May 2025 16:44:09 GMT
< server: uvicorn
< content-length: 136
< content-type: application/json
< 
* Connection #0 to host 127.0.0.1 left intact
{"object":"error","message":"JSON decode error: Expecting value: line 1 column 1 (char 0)","type":"Bad Request","param":null,"code":400}
```
</details>

davidxia committed:
[Bugfix]: make most of `test_openai_schema.py` pass

This change makes most of the parameterized tests in `pytest
tests/entrypoints/openai/test_openai_schema.py` pass. We handle user input
errors, enrich HTTP error messages with more information, and update the error
response definitions for several API endpoints.

The remaining tests that fail are `POST /tokenize` and maybe `POST
/v1/chat/completions`.

FIX #17037
FIX #17038

Signed-off-by: David Xia <david@davidxia.com>

davidxia commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -735,12 +816,30 @@ async def is_sleeping(raw_request: Request):
+                 status.HTTP_400_BAD_REQUEST: {
+                     "model": ErrorResponse
+                 },
+                 status.HTTP_415_UNSUPPORTED_MEDIA_TYPE: {
+                     "model": HTTPError
+                 },
```
otherwise fails with

<details>
```
________________________________________________ test_openapi_stateless (verbose_name='POST /invocations') ________________________________________________

    @wraps(test)
>   def test_function(*args: Any, **kwargs: Any) -> Any:

.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:83: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:160: in async_run
    loop.run_until_complete(future)
/usr/lib/python3.12/asyncio/base_events.py:687: in run_until_complete
    return future.result()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

case = Case()

    @schema.parametrize()
    @schema.override(headers={"Content-Type": "application/json"})
    async def test_openapi_stateless(case):
        #No need to verify SSL certificate for localhost
>       await case.call_and_validate(verify=False)
E       schemathesis.exceptions.CheckFailed: 
E       
E       1. Undocumented HTTP status code
E       
E           Received: 415
E           Documented: 200
E       
E       [415] Unsupported Media Type:
E       
E           `{"detail":"Unsupported Media Type: Only 'application/json' is allowed"}`
E       
E       Reproduce with: 
E       
E           curl -X POST --insecure http://localhost:52711/invocations
E       
E       Falsifying example: async_run(
E           case=,
E       )

tests/entrypoints/openai/test_openai_schema.py:49: CheckFailed
```
</details>

davidxia commented on tests/entrypoints/openai/test_openai_schema.py:
```diff
@@ -45,5 +45,5 @@ def get_schema(server):
 @schema.override(headers={"Content-Type": "application/json"})
 async def test_openapi_stateless(case):
-    #No need to verify SSL certificate for localhost
-    await case.call_and_validate(verify=False)
```
Remove `await`, otherwise many test cases fail like so. [Docs also suggest we don't need](https://schemathesis.readthedocs.io/en/stable/python.html#async-support) `await`.

<details>
```
_________________________________________________ test_openapi_stateless (verbose_name='POST /v2/rerank') _________________________________________________

    @wraps(test)
>   def test_function(*args: Any, **kwargs: Any) -> Any:

.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:83: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:160: in async_run
    loop.run_until_complete(future)
/usr/lib/python3.12/asyncio/base_events.py:687: in run_until_complete
    return future.result()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

case = Case(body={'documents': [], 'query': ''})

    @schema.parametrize()
    @schema.override(headers={"Content-Type": "application/json"})
    async def test_openapi_stateless(case):
        #No need to verify SSL certificate for localhost
>       await case.call_and_validate(verify=False)
E       TypeError: object Response can't be used in 'await' expression
E       Falsifying example: async_run(
E           case=,
E       )

tests/entrypoints/openai/test_openai_schema.py:49: TypeError
```
</details>

davidxia commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -601,12 +673,21 @@ async def create_score_v1(request: ScoreRequest, raw_request: Request):
-@router.post("/v1/audio/transcriptions")
+@router.post("/v1/audio/transcriptions",
+             responses={
+                 status.HTTP_400_BAD_REQUEST: {
+                     "model": ErrorResponse
+                 },
```
fixes errors like

<details>
```
__________________________________________ test_openapi_stateless (verbose_name='POST /v1/audio/transcriptions') __________________________________________

    @wraps(test)
>   def test_function(*args: Any, **kwargs: Any) -> Any:

.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:83: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:160: in async_run
    loop.run_until_complete(future)
/usr/lib/python3.12/asyncio/base_events.py:687: in run_until_complete
    return future.result()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

case = Case(body={'file': Binary(data=b'')})

    @schema.parametrize()
    @schema.override(headers={"Content-Type": "application/json"})
    async def test_openapi_stateless(case):
        #No need to verify SSL certificate for localhost
>       await case.call_and_validate(verify=False)
E       schemathesis.exceptions.CheckFailed: 
E       
E       1. Undocumented HTTP status code
E       
E           Received: 400
E           Documented: 200, 422
E       
E       [400] Bad Request:
E       
E           `{"object":"error","message":"[{'type': 'value_error', 'loc': ('body', 'file'), 'msg': \"Value error, Expected UploadFile, received: <class 'str'>\", 'input': \"b''\", 'ctx': {'error': ValueError(\"Expected UploadFile, received: <class 'str'>\")}}]","type":"BadRequestError","param":null,"code":400}`
E       
E       Reproduce with: 
E       
E           curl -X POST -H 'Content-Type: application/x-www-form-urlencoded' -d file=b%27%27 --insecure http://localhost:52711/v1/audio/transcriptions
E       
E       Falsifying example: async_run(
E           case=,
E       )

tests/entrypoints/openai/test_openai_schema.py:49: CheckFailed
```
</details>

davidxia commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -486,7 +536,19 @@ async def create_chat_completion(request: ChatCompletionRequest,
+@router.post("/v1/completions",
+             dependencies=[Depends(validate_json_request)],
+             responses={
+                 status.HTTP_400_BAD_REQUEST: {
+                     "model": ErrorResponse
+                 },
```
fixes errors like

<details>
```
______________________________________________ test_openapi_stateless (verbose_name='POST /v1/completions') _______________________________________________

    @wraps(test)
>   def test_function(*args: Any, **kwargs: Any) -> Any:

.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:83: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:160: in async_run
    loop.run_until_complete(future)
/usr/lib/python3.12/asyncio/base_events.py:687: in run_until_complete
    return future.result()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

case = Case(body={'prompt': []})

    @schema.parametrize()
    @schema.override(headers={"Content-Type": "application/json"})
    async def test_openapi_stateless(case):
        #No need to verify SSL certificate for localhost
>       await case.call_and_validate(verify=False)
E       schemathesis.exceptions.CheckFailed: 
E       
E       1. Undocumented HTTP status code
E       
E           Received: 400
E           Documented: 200, 422
E       
E       [400] Bad Request:
E       
E           `{"object":"error","message":"please provide at least one prompt","type":"BadRequestError","param":null,"code":400}`
E       
E       Reproduce with: 
E       
E           curl -X POST -H 'Content-Type: application/json' -d '{"prompt": []}' --insecure http://localhost:52711/v1/completions
E       
E       Falsifying example: async_run(
E           case=,
E       )

tests/entrypoints/openai/test_openai_schema.py:49: CheckFailed
```
</details>

davidxia commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -464,7 +506,15 @@ async def show_version():
-             dependencies=[Depends(validate_json_request)])
+             dependencies=[Depends(validate_json_request)],
+             responses={
+                 status.HTTP_400_BAD_REQUEST: {
+                     "model": ErrorResponse
+                 },
```
fixes errors like

<details>
```
____________________________________________ test_openapi_stateless (verbose_name='POST /v1/chat/completions') ____________________________________________

    @wraps(test)
>   def test_function(*args: Any, **kwargs: Any) -> Any:

.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:83: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:160: in async_run
    loop.run_until_complete(future)
/usr/lib/python3.12/asyncio/base_events.py:687: in run_until_complete
    return future.result()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

case = Case(body={'messages': []})

    @schema.parametrize()
    @schema.override(headers={"Content-Type": "application/json"})
    async def test_openapi_stateless(case):
        #No need to verify SSL certificate for localhost
>       await case.call_and_validate(verify=False)
E       schemathesis.exceptions.CheckFailed: 
E       
E       1. Undocumented HTTP status code
E       
E           Received: 400
E           Documented: 200, 422
E       
E       [400] Bad Request:
E       
E           `{"object":"error","message":"","type":"BadRequestError","param":null,"code":400}`
E       
E       Reproduce with: 
E       
E           curl -X POST -H 'Content-Type: application/json' -d '{"messages": []}' --insecure http://localhost:52711/v1/chat/completions
E       
E       Falsifying example: async_run(
E           case=,
E       )

tests/entrypoints/openai/test_openai_schema.py:49: CheckFailed
```
</details>

davidxia commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -413,13 +420,26 @@ async def get_server_load_metrics(request: Request):
+@router.post("/tokenize",
+             dependencies=[Depends(validate_json_request)],
+             responses={
+                 status.HTTP_400_BAD_REQUEST: {
+                     "model": ErrorResponse
+                 },
```
fixes errors like

<details>
```
_________________________________________________ test_openapi_stateless (verbose_name='POST /tokenize') __________________________________________________

    @wraps(test)
>   def test_function(*args: Any, **kwargs: Any) -> Any:

.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:83: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:160: in async_run
    loop.run_until_complete(future)
/usr/lib/python3.12/asyncio/base_events.py:687: in run_until_complete
    return future.result()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

case = Case(body={'messages': []})

    @schema.parametrize()
    @schema.override(headers={"Content-Type": "application/json"})
    async def test_openapi_stateless(case):
        #No need to verify SSL certificate for localhost
>       await case.call_and_validate(verify=False)
E       schemathesis.exceptions.CheckFailed: 
E       
E       1. Undocumented HTTP status code
E       
E           Received: 400
E           Documented: 200, 422
E       
E       [400] Bad Request:
E       
E           `{"object":"error","message":"","type":"BadRequestError","param":null,"code":400}`
E       
E       Reproduce with: 
E       
E           curl -X POST -H 'Content-Type: application/json' -d '{"messages": []}' --insecure http://localhost:52711/tokenize
E       
E       Falsifying example: async_run(
E           case=,
E       )

tests/entrypoints/openai/test_openai_schema.py:49: CheckFailed
```
</details>

davidxia commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -413,13 +420,26 @@ async def get_server_load_metrics(request: Request):
 
 
-@router.api_route("/ping", methods=["GET", "POST"])
+@router.get("/ping", response_class=Response)
+@router.post("/ping", response_class=Response)
```
`@router.api_route` isn't recommended. See https://github.com/fastapi/fastapi/discussions/7736#discussioncomment-5143904

`response_class=Response` fixes errors

<details>
```
___________________________________________________ test_openapi_stateless (verbose_name='POST /ping') ____________________________________________________

    @wraps(test)
>   def test_function(*args: Any, **kwargs: Any) -> Any:

.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:83: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:160: in async_run
    loop.run_until_complete(future)
/usr/lib/python3.12/asyncio/base_events.py:687: in run_until_complete
    return future.result()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

case = Case()

    @schema.parametrize()
    @schema.override(headers={"Content-Type": "application/json"})
    async def test_openapi_stateless(case):
        #No need to verify SSL certificate for localhost
>       await case.call_and_validate(verify=False)
E       schemathesis.exceptions.CheckFailed: 
E       
E       1. Missing Content-Type header
E       
E           The following media types are documented in the schema:
E           - `application/json`
E       
E       [200] OK:
E       
E           <EMPTY>
E       
E       Reproduce with: 
E       
E           curl -X POST --insecure http://localhost:52711/ping
E       
E       Falsifying example: async_run(
E           case=,
E       )

tests/entrypoints/openai/test_openai_schema.py:49: CheckFailed
```
</details>

davidxia commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -388,7 +395,7 @@ def engine_client(request: Request) -> EngineClient:
 
 
-@router.get("/health")
+@router.get("/health", response_class=Response)
```
fixes errors like

<details>
```
___________________________________________________ test_openapi_stateless (verbose_name='GET /health') ___________________________________________________

    @wraps(test)
>   def test_function(*args: Any, **kwargs: Any) -> Any:

.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:83: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:160: in async_run
    loop.run_until_complete(future)
/usr/lib/python3.12/asyncio/base_events.py:687: in run_until_complete
    return future.result()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

case = Case()

    @schema.parametrize()
    @schema.override(headers={"Content-Type": "application/json"})
    async def test_openapi_stateless(case):
        #No need to verify SSL certificate for localhost
>       await case.call_and_validate(verify=False)
E       schemathesis.exceptions.CheckFailed: 
E       
E       1. Missing Content-Type header
E       
E           The following media types are documented in the schema:
E           - `application/json`
E       
E       [200] OK:
E       
E           <EMPTY>
E       
E       Reproduce with: 
E       
E           curl -X GET --insecure http://localhost:52711/health
E       
E       Falsifying example: async_run(
E           case=,
E       )

tests/entrypoints/openai/test_openai_schema.py:49: CheckFailed
```
</details>

davidxia commented on .buildkite/test-pipeline.yaml:
```diff
@@ -121,7 +121,7 @@ steps:
   - pytest -v -s entrypoints/llm/test_generate_multiple_loras.py # it needs a clean process
   - VLLM_USE_V1=0 pytest -v -s entrypoints/llm/test_guided_generate.py # it needs a clean process
-  - pytest -v -s entrypoints/openai --ignore=entrypoints/openai/test_oot_registration.py  --ignore=entrypoints/openai/test_chat_with_tool_reasoning.py --ignore=entrypoints/openai/correctness/ --ignore=entrypoints/openai/test_openai_schema.py
+  - pytest -v -s entrypoints/openai --ignore=entrypoints/openai/test_oot_registration.py  --ignore=entrypoints/openai/test_chat_with_tool_reasoning.py --ignore=entrypoints/openai/correctness/
```
are we ready to enable the test?

tarukumar replied:
I would suggest to ignore it until the test become more stable. Otherwise automrge would be not useful due to the failure of the test.

davidxia commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -296,10 +298,9 @@ async def validate_json_request(raw_request: Request):
-            status_code=HTTPStatus.UNSUPPORTED_MEDIA_TYPE,
-            detail="Unsupported Media Type: Only 'application/json' is allowed"
-        )
+        raise RequestValidationError(errors=[
+            "Unsupported Media Type: Only 'application/json' is allowed"
+        ])
```
We raise a `RequestValidationError` so that the existing `validation_exception_handler()` will convert it to a `JSONResponse`. We need that to happen so we can specify `ErrorResponse` as the return type in API endpoints like `@router.post("/invocations", ...)`. FastAPI requires return types to be Pydantic types. `ErrorResponse` inherits from Pydantic's `BaseModel`.

Otherwise fails with errors like

<details>

```
__________________________________________________________________________ test_openapi_stateless (verbose_name='POST /v1/completions') ___________________________________________________________________________

    @wraps(test)
>   def test_function(*args: Any, **kwargs: Any) -> Any:

.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:83: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:160: in async_run
    loop.run_until_complete(future)
/usr/lib/python3.12/asyncio/base_events.py:687: in run_until_complete
    return future.result()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

case = Case(body={'prompt': [], 'logprobs': -1})

    @schema.parametrize()
    @schema.override(headers={"Content-Type": "application/json"})
    async def test_openapi_stateless(case):
        # No need to verify SSL certificate for localhost
>       case.call_and_validate(verify=False)
E       schemathesis.exceptions.CheckFailed: 
E       
E       1. Server error
E       
E       2. Undocumented Content-Type
E       
E           Received: text/plain; charset=utf-8
E           Documented: application/json
E       
E       [500] Internal Server Error:
E       
E           `Internal Server Error`
E       
E       Reproduce with: 
E       
E           curl -X POST -H 'Content-Type: application/json' -d '{"prompt": [], "logprobs": -1}' --insecure http://localhost:45731/v1/completions
E       
E       Falsifying example: async_run(
E           case=,
E       )

tests/entrypoints/openai/test_openai_schema.py:49: CheckFailed

____________________________________________________________________________ test_openapi_stateless (verbose_name='POST /invocations') ____________________________________________________________________________

    @wraps(test)
>   def test_function(*args: Any, **kwargs: Any) -> Any:

.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:83: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:160: in async_run
    loop.run_until_complete(future)
/usr/lib/python3.12/asyncio/base_events.py:687: in run_until_complete
    return future.result()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

case = Case()

    @schema.parametrize()
    @schema.override(headers={"Content-Type": "application/json"})
    async def test_openapi_stateless(case):
        # No need to verify SSL certificate for localhost
>       case.call_and_validate(verify=False)
E       schemathesis.exceptions.CheckFailed: 
E       
E       1. Response violates schema
E       
E           'message' is a required property
E       
E           Schema:
E       
E               {
E                   "properties": {
E                       "object": {
E                           "type": "string",
E                           "title": "Object",
E                           "default": "error"
E                       },
E                       "message": {
E                           "type": "string",
E                           "title": "Message"
E                       },
E                       "type": {
E                           "type": "string",
E                           "title": "Type"
E                       },
E                       "param": {
E                           "anyOf": [
E                               {
E                                   "type": "string"
E                   // Output truncated...
E               }
E       
E           Value:
E       
E               {
E                   "detail": "Unsupported Media Type: Only 'application/json' is allowed"
E               }
E       
E       [415] Unsupported Media Type:
E       
E           `{"detail":"Unsupported Media Type: Only 'application/json' is allowed"}`
E       
E       Reproduce with: 
E       
E           curl -X POST --insecure http://localhost:45731/invocations
E       
E       Falsifying example: async_run(
E           case=,
E       )

tests/entrypoints/openai/test_openai_schema.py:49: CheckFailed
```
</details>

davidxia commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -464,7 +505,23 @@ async def show_version():
-             dependencies=[Depends(validate_json_request)])
+             dependencies=[Depends(validate_json_request)],
+             responses={
+                 status.HTTP_200_OK: {
+                     "content": {
+                         "text/event-stream": {}
+                     }
+                 },
```
Needed to test expecting `Content-Type: application/json` all the time. See https://fastapi.tiangolo.com/advanced/additional-responses/#additional-media-types-for-the-main-response

davidxia commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -333,7 +339,7 @@ def mount_metrics(app: FastAPI):
         ],
         registry=registry,
-    ).add().instrument(app).expose(app)
+    ).add().instrument(app).expose(app, response_class=PrometheusResponse)
```
see https://github.com/trallnag/prometheus-fastapi-instrumentator/issues/163#issue-1296092364

davidxia replied:
yup, done!

davidxia commented on tests/entrypoints/openai/test_openai_schema.py:
```diff
@@ -45,5 +48,13 @@ def get_schema(server):
+        case.operation.path,
+    )
+    timeout = {
+        ("POST", "/v1/chat/completions"): 99999,
```
this endpoint often times out for me. See https://schemathesis.readthedocs.io/en/stable/recipes.html#per-route-request-timeouts

davidxia commented on tests/entrypoints/openai/test_openai_schema.py:
```diff
@@ -45,5 +48,13 @@ def get_schema(server):
+        case.operation.path,
+    )
+    timeout = {
+        ("POST", "/v1/chat/completions"): 30,
```
this endpoint often times out for me. See https://schemathesis.readthedocs.io/en/stable/recipes.html#per-route-request-timeouts

tarukumar replied:
Odd for me it didn't happen

davidxia replied:
might be because I'm using CPU platform. I'll revert.

tarukumar commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -419,7 +421,16 @@ async def ping(raw_request: Request) -> Response:
+                 500: {
+                     "model": ErrorResponse
+                 },
+             })
```
I think so we should do that. I am open to hear others opinion as well

DarkLight1337 replied:
Yeah that sounds good

davidxia replied:
updated

hmellor reviewed (commented):
This looks nice so far, although I'm not sure it solves https://github.com/vllm-project/vllm/issues/17038 as that issue says that some endpoints are returning 500 where they should be returning 422.

Also, should we stick either to `http.HTTPStatus` or `fastapi.status`? It's a little confusing to have both (I'm slightly leaning towards the `fastapi` ones because they include the number and the name).

hmellor commented:
FYI I've made https://github.com/vllm-project/vllm/pull/17909 to change the error code for transcription

hmellor commented:
> I can update to only use http.HTTPStatus to be consistent throughout the module. Wdyt?

This is a good point actually! Could you also reduce reprtition by allowing the full `HTTPStatus.XYZ` to be passed to `create_error_code` and then the message, type and code (if not explicitly specified) can be extracted from the `HTTPStatus`?

davidxia commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -474,7 +511,23 @@ async def show_version():
-             dependencies=[Depends(validate_json_request)])
+             dependencies=[Depends(validate_json_request)],
+             responses={
+                 HTTPStatus.OK.value: {
+                     "content": {
+                         "text/event-stream": {}
+                     }
```
Needed to test expecting `Content-Type: application/json` all the time. See https://fastapi.tiangolo.com/advanced/additional-responses/#additional-media-types-for-the-main-response

davidxia commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -397,11 +402,11 @@ def engine_client(request: Request) -> EngineClient:
-async def health(raw_request: Request) -> JSONResponse:
+@router.get("/health", response_class=Response)
+async def health(raw_request: Request) -> Response:
     """Health check."""
     await engine_client(raw_request).check_health()
-    return JSONResponse(content={}, status_code=200)
+    return Response(status_code=200)
```
changed to return a `Response` since there's no JSON payload to return

davidxia commented:
> This looks nice so far, although I'm not sure it solves #17038 as that issue says that some endpoints are returning 500 where they should be returning 422.

All the requests that receive unexpected 500 responses documented in #17038 now receive 400. I think this PR now fixes that issue.

<details>
```
$ curl -X POST -H 'Content-Type: application/json' -d '{"messages": []}' --insecure http://127.0.0.1:8000/tokenize -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /tokenize HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 16
> 
< HTTP/1.1 400 Bad Request
< date: Mon, 12 May 2025 12:54:20 GMT
< server: uvicorn
< content-length: 104
< content-type: application/json
< 
* Connection #0 to host 127.0.0.1 left intact
{"object":"error","message":" list index out of range","type":"BadRequestError","param":null,"code":400}





$ curl -X POST -H 'Content-Type: application/json' -d '{"tokens": [-1]}' --insecure http://127.0.0.1:8000/detokenize -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /detokenize HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 16
> 
< HTTP/1.1 400 Bad Request
< date: Mon, 12 May 2025 13:26:25 GMT
< server: uvicorn
< content-length: 128
< content-type: application/json
< 
* Connection #0 to host 127.0.0.1 left intact
{"object":"error","message":" ['out of range integral type conversion attempted']","type":"Bad Request","param":null,"code":400}


$ curl -X POST -H 'Content-Type: application/json' -d '{"messages": []}' --insecure http://127.0.0.1:8000/v1/chat/completions -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /v1/chat/completions HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 16
> 
< HTTP/1.1 400 Bad Request
< date: Mon, 12 May 2025 13:26:58 GMT
< server: uvicorn
< content-length: 104
< content-type: application/json
< 
* Connection #0 to host 127.0.0.1 left intact
{"object":"error","message":" list index out of range","type":"BadRequestError","param":null,"code":400}


$ curl -X POST -H 'Content-Type: application/json' -d '{"prompt": [-1]}' --insecure http://127.0.0.1:8000/v1/completions -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /v1/completions HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 16
> 
< HTTP/1.1 400 Bad Request
< date: Mon, 12 May 2025 13:30:40 GMT
< server: uvicorn
< content-length: 123
< content-type: application/json
< 
* Connection #0 to host 127.0.0.1 left intact
{"object":"error","message":"out of range integral type conversion attempted","type":"Bad Request","param":null,"code":400}



$ curl -X POST -H 'Content-Type: application/json' --insecure http://127.0.0.1:8000/invocations -v
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> POST /invocations HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> 
< HTTP/1.1 400 Bad Request
< date: Mon, 12 May 2025 13:31:35 GMT
< server: uvicorn
< content-length: 136
< content-type: application/json
< 
* Connection #0 to host 127.0.0.1 left intact
{"object":"error","message":"JSON decode error: Expecting value: line 1 column 1 (char 0)","type":"Bad Request","param":null,"code":400}
```
</details>

hmellor commented:
> All the requests that receive unexpected 500 responses documented in https://github.com/vllm-project/vllm/issues/17038 now receive 400. I think this PR now fixes that issue.

Shouldn't they recieve 422? That's what the linked issue seems to suggest

davidxia commented:
> This is a good point actually! Could you also reduce reprtition by allowing the full `HTTPStatus.XYZ` to be passed to `create_error_code` and then the message, type and code (if not explicitly specified) can be extracted from the `HTTPStatus`?

I actually just replaced almost all the `return ErrorResponse()` in the endpoints with `raise HTTPException()` because the former returned a 200 OK response instead of the expected HTTP error codes. Ready for another review.

davidxia commented:
> > All the requests that receive unexpected 500 responses documented in #17038 now receive 400. I think this PR now fixes that issue.
> 
> Shouldn't they recieve 422? That's what the linked issue seems to suggest

The error in that issue is saying that only 200 and 422 are documented. I've added more expected codes like 400. HTTP 422 specifically points to semantic or validation errors with the request payload, HTTP 400 broadly indicates issues with the request syntax or structure. Happy to change to whatever makes more sense. lmk!

davidxia commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -862,10 +1036,26 @@ def build_app(args: Namespace) -> FastAPI:
         allow_headers=args.allowed_headers,
     )
 
+    @app.exception_handler(HTTPException)
+    async def http_exception_handler(_: Request, exc: HTTPException):
+        err = ErrorResponse(message=exc.detail,
+                            type=HTTPStatus(exc.status_code).phrase,
+                            code=exc.status_code)
+        return JSONResponse(err.model_dump(), status_code=exc.status_code)
```
This is needed to fix errors like the below where the FastAPI middleware returns an error before our custom endpoint function runs. We need to return a payload that has the schema of `ErrorResponse` because we document that as the return type in our endpoints.

<details>

```
______________________________________________________________ test_openapi_stateless (verbose_name='POST /v1/audio/transcriptions') ______________________________________________________________

    @wraps(test)
>   def test_function(*args: Any, **kwargs: Any) -> Any:

.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:83: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

case = Case(body={'file': Binary(data=b'')})

    @schema.parametrize()
    @schema.override(headers={"Content-Type": "application/json"})
    def test_openapi_stateless(case: schemathesis.Case):
        key = (
            case.operation.method.upper(),
            case.operation.path,
        )
        timeout = {
            ("POST", "/v1/chat/completions"): 99999,
        }.get(key, 10)
    
        #No need to verify SSL certificate for localhost
>       case.call_and_validate(verify=False, timeout=timeout)
E       schemathesis.exceptions.CheckFailed: 
E       
E       1. Response violates schema
E       
E           'message' is a required property
E       
E           Schema:
E       
E               {
E                   "properties": {
E                       "object": {
E                           "type": "string",
E                           "title": "Object",
E                           "default": "error"
E                       },
E                       "message": {
E                           "type": "string",
E                           "title": "Message"
E                       },
E                       "type": {
E                           "type": "string",
E                           "title": "Type"
E                       },
E                       "param": {
E                           "anyOf": [
E                               {
E                                   "type": "string"
E                   // Output truncated...
E               }
E       
E           Value:
E       
E               {
E                   "detail": "Expected 'file' to be a file-like object, not 'str'."
E               }
E       
E       [422] Unprocessable Entity:
E       
E           `{"detail":"Expected 'file' to be a file-like object, not 'str'."}`
E       
E       Reproduce with: 
E       
E           curl -X POST -H 'Content-Type: application/x-www-form-urlencoded' -d file=b%27%27 --insecure http://localhost:36871/v1/audio/transcriptions
E       
E       Falsifying example: test_openapi_stateless(
E           case=,
E       )

tests/entrypoints/openai/test_openai_schema.py:57: CheckFailed
```
</details>

hmellor reviewed (approved):
This is an exceptionally well documented PR!

I only have one tiny nit about including some of the context you provided on GitHub into code comments

davidxia reviewed (commented):
@hmellor thank you. I updated the PR description after rebasing. Ready for another review. 🙏

davidxia commented:
The only test that consistently fails for me now is `tests/entrypoints/openai/test_openai_schema.py::test_openapi_stateless[POST /tokenize]` caused by https://github.com/vllm-project/vllm/blob/98ea35601cdb34fdd618f965e7bcc3cb02a677fc/vllm/entrypoints/chat_utils.py#L1083 when `part_type` is `"file"`.

<details>

```
============================================================================================ FAILURES =============================================================================================
_____________________________________________________________________ test_openapi_stateless (verbose_name='POST /tokenize') ______________________________________________________________________

    @wraps(test)
>   def test_function(*args: Any, **kwargs: Any) -> Any:

.venv/lib/python3.12/site-packages/schemathesis/_hypothesis.py:83: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

case = Case(body={'messages': [{'content': [{'file': {}, 'type': 'file'}], 'role': 'user'}]})

    @schema.parametrize(tag="debug")
    @schema.override(headers={"Content-Type": "application/json"})
    def test_openapi_stateless(case: schemathesis.Case):
        key = (
            case.operation.method.upper(),
            case.operation.path,
        )
        timeout = {
            ("POST", "/v1/chat/completions"): 99999,
        }.get(key, 10)
    
        #No need to verify SSL certificate for localhost
>       case.call_and_validate(verify=False, timeout=timeout)
E       schemathesis.exceptions.CheckFailed: 
E       
E       1. Server error
E       
E       [500] Internal Server Error:
E       
E           `{"object":"error","message":"Unknown part type: file","type":"Internal Server Error","param":null,"code":500}`
E       
E       Reproduce with: 
E       
E           curl -X POST -H 'Content-Type: application/json' -d '{"messages": [{"content": [{"file": {}, "type": "file"}], "role": "user"}]}' --insecure http://localhost:48767/tokenize
E       
E       Falsifying example: test_openapi_stateless(
E           case=,
E       )

tests/entrypoints/openai/test_openai_schema.py:57: CheckFailed
```

```bash
$ curl -X POST -H 'Content-Type: application/json' -d '{"messages": [{"content": [{"file": {}, "type": "file"}], "role": "user"}]}' --insecure http://localhost:8000/tokenize -v
Note: Unnecessary use of -X or --request, POST is already inferred.
* Host localhost:8000 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:8000...
* connect to ::1 port 8000 from ::1 port 40420 failed: Connection refused
*   Trying 127.0.0.1:8000...
* Connected to localhost (127.0.0.1) port 8000
> POST /tokenize HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 75
> 
< HTTP/1.1 500 Internal Server Error
< date: Mon, 12 May 2025 17:32:11 GMT
< server: uvicorn
< content-length: 109
< content-type: application/json
< 
* Connection #0 to host localhost left intact
{"object":"error","message":"Unknown part type: file","type":"Internal Server Error","param":null,"code":500}
```
</details>

davidxia commented:
This is ready for review 🙏

davidxia commented:
I rebased on main and updated `POST /tokenize` to return a [501 Not Implemented](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/501) instead of 500 if a Python `NotImplemented` exception is raised. It doesn't fix the test because according to the OpenAPI spec, `$ curl -X POST -H 'Content-Type: application/json' -d '{"messages": [{"content": [{"file": {}, "type": "file"}], "role": "user"}]}' --insecure http://localhost:8000/tokenize` is a valid request that should get a 200 reply. But at least the reply status code is more specific.

We can discuss how to make the test case pass in a follow-up PR and then enabling the overall test by removing the `--ignore=entrypoints/openai/test_openai_schema.py` from `test-pipeline.yaml`. But I think this PR is definitely an improvement over the status quo.

davidxia commented:
I created a follow-up issue https://github.com/vllm-project/vllm/issues/18162.
